python/tests/test_integration_schema.py

```diff
@@ -0,0 +1,2617 @@
+    assert "already exists: location.latitude" in str(exc_info.value)
+
+
+@pytest.mark.integration
+def test_add_to_non_struct_type(catalog: Catalog, table_schema_simple: Schema) -> None:
+    table = _create_table_with_schema(catalog, table_schema_simple)
+    update = UpdateSchema(table)
+    with pytest.raises(ValueError) as exc_info:
+        update.add_column(path=("foo", "lat"), field_type=IntegerType())
+    assert "Cannot add column 'lat' to non-struct type" in str(exc_info.value)
+
+
+@pytest.mark.integration
+@pytest.mark.integration
+def test_schema_evolution_via_transaction(catalog: Catalog) -> None:
+    schema = Schema(
+        NestedField(field_id=1, name="col_uuid", field_type=UUIDType(), required=False),
+        NestedField(field_id=2, name="col_fixed", field_type=FixedType(25), required=False),
+    )
+    tbl = _create_table_with_schema(catalog, schema)
+
+    assert tbl.schema() == schema
+
+    with tbl.transaction() as tx:
+        tx.update_schema().add_column("col_string", StringType()).commit()
+
+    assert tbl.schema() == Schema(
+        NestedField(field_id=1, name="col_uuid", field_type=UUIDType(), required=False),
+        NestedField(field_id=2, name="col_fixed", field_type=FixedType(25), required=False),
+        NestedField(field_id=3, name="col_string", field_type=StringType(), required=False),
+        schema_id=1,
+    )
+
+    tbl.update_schema().add_column("col_integer", IntegerType()).commit()
+
+    assert tbl.schema() == Schema(
+        NestedField(field_id=1, name="col_uuid", field_type=UUIDType(), required=False),
+        NestedField(field_id=2, name="col_fixed", field_type=FixedType(25), required=False),
+        NestedField(field_id=3, name="col_string", field_type=StringType(), required=False),
+        NestedField(field_id=4, name="col_integer", field_type=IntegerType(), required=False),
+        schema_id=1,
+    )
+
+    with pytest.raises(CommitFailedException) as exc_info:
+        with tbl.transaction() as tx:
+            # Start a new update
+            schema_update = tx.update_schema()
+
+            # Do a concurrent update
+            tbl.update_schema().add_column("col_long", LongType()).commit()
+
+            # stage another update in the transaction
+            schema_update.add_column("col_double", DoubleType()).commit()
+
+    assert "Requirement failed: current schema changed: expected id 2 != 3" in str(exc_info.value)
+
+    assert tbl.schema() == Schema(
+        NestedField(field_id=1, name="col_uuid", field_type=UUIDType(), required=False),
+        NestedField(field_id=2, name="col_fixed", field_type=FixedType(25), required=False),
+        NestedField(field_id=3, name="col_string", field_type=StringType(), required=False),
+        NestedField(field_id=4, name="col_integer", field_type=IntegerType(), required=False),
+        NestedField(field_id=5, name="col_long", field_type=LongType(), required=False),
+        schema_id=1,
```

Comment:
Each schema change should produce a new schema ID in table metadata. If you were testing just `_apply` then this would be correct, but since this is actually committed, the schema ID needs to be assigned for every change.

@Fokko, this is a blocker.

Reply:
The schema-IDs are updated correctly, but we don't validate them. The Python `__eq__` is the same as Java's schema: https://github.com/apache/iceberg/blob/master/api/src/main/java/org/apache/iceberg/Schema.java#L464-L473

Reply:
![Uploading image.png…]()